### PR TITLE
Saucelabs in travis refactored

### PR DESCRIPTION
**Pull request by Droogans**
Putting back saucelabs support for testing. Consider this to be a continuation of the (failed) trial from earlier.

This should be merged only once we can prove that we have documentation and support for continued usage on SauceLabs. We shouldn't add it to the team's workflow unless everyone understands what is it, where it lives, what it does, and how to see it.

Ideally, most of the setup will occur during the initial `yarn install` command in the `test/` directory.

### LGTM's
- [x] Dev LGTM
- [x] Zoom LGTM

### CODE REVIEW

File: test/dom-snapshots/components-typography-caption-MicrosoftEdge.js.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Snapshot report for `built/dom-snapshots/components-typography-caption-MicrosoftEdge.js`
```
Comment (CITguy) [inline test/dom-snapshots/components-typography-caption-MicrosoftEdge.js.md]: Please remove. Old artifact from visreg testing.

File: test/dom-snapshots/components-typography-caption-chrome.js.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Snapshot report for `built/dom-snapshots/components-typography-caption-chrome.js`
```
Comment (CITguy) [inline test/dom-snapshots/components-typography-caption-chrome.js.md]: Please remove. Old artifact from visreg testing.

File: test/dom-snapshots/components-typography-caption-firefox.js.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# Snapshot report for `built/dom-snapshots/components-typography-caption-firefox.js`
```
Comment (CITguy) [inline test/dom-snapshots/components-typography-caption-firefox.js.md]: Please remove. Old artifact from visreg testing.

File: test/dom-snapshots/components-typography-caption-internet-explorer.js.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Snapshot report for `built/dom-snapshots/components-typography-caption-internet-explorer.js`
```
Comment (CITguy) [inline test/dom-snapshots/components-typography-caption-internet-explorer.js.md]: Please remove. Old artifact from visreg testing.

File: test/dom-snapshots/components-typography-caption-safari.js.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# Snapshot report for `built/dom-snapshots/components-typography-caption-safari.js`
```
Comment (CITguy) [inline test/dom-snapshots/components-typography-caption-safari.js.md]: Please remove. Old artifact from visreg testing.

File: test/dom-snapshots/components-typography-error-message-MicrosoftEdge.js.md [hunk 6]
```diff
@@ -0,0 +1,13 @@
+# Snapshot report for `built/dom-snapshots/components-typography-error-message-MicrosoftEdge.js`
```
Comment (CITguy) [inline test/dom-snapshots/components-typography-error-message-MicrosoftEdge.js.md]: Please remove. Old artifact from visreg testing.

File: test/dom-snapshots/components-typography-error-message-chrome.js.md [hunk 7]
```diff
@@ -0,0 +1,13 @@
+# Snapshot report for `built/dom-snapshots/components-typography-error-message-chrome.js`
```
Comment (CITguy) [inline test/dom-snapshots/components-typography-error-message-chrome.js.md]: Please remove. Old artifact from visreg testing.

File: test/dom-snapshots/components-typography-error-message-firefox.js.md [hunk 8]
```diff
@@ -0,0 +1,13 @@
+# Snapshot report for `built/dom-snapshots/components-typography-error-message-firefox.js`
```
Comment (CITguy) [inline test/dom-snapshots/components-typography-error-message-firefox.js.md]: Please remove. Old artifact from visreg testing.

File: test/dom-snapshots/components-typography-error-message-internet-explorer.js.md [hunk 9]
```diff
@@ -0,0 +1,13 @@
+# Snapshot report for `built/dom-snapshots/components-typography-error-message-internet-explorer.js`
```
Comment (CITguy) [inline test/dom-snapshots/components-typography-error-message-internet-explorer.js.md]: Please remove. Old artifact from visreg testing.

File: test/dom-snapshots/components-typography-error-message-safari.js.md [hunk 10]
```diff
@@ -0,0 +1,13 @@
+# Snapshot report for `built/dom-snapshots/components-typography-error-message-safari.js`
```
Comment (CITguy) [inline test/dom-snapshots/components-typography-error-message-safari.js.md]: Please remove. Old artifact from visreg testing.

File: test/regression/runSingleTest.ts [hunk 11]
```diff
@@ -3,22 +3,88 @@ import {By, snap, Snappit, IConfig, WebDriver, WebElement} from "snappit-visual-
...
+        safari: MAC,
+        "internet explorer": "Windows 10",
+        MicrosoftEdge: "Windows 10"
+    };
+    const platform = platforms[browser];
+
     const config: IConfig = {
-        browser: browser
+        browser: browser,
+        screenshotsDir: "artifacts/regressionScreenshots",
```
Comment (CITguy) [inline test/regression/runSingleTest.ts]: Do we need this if we're not taking screenshots?

File: test/regression/runSingleTest.ts [hunk 12]
```diff
@@ -3,22 +3,88 @@ import {By, snap, Snappit, IConfig, WebDriver, WebElement} from "snappit-visual-
...
+            "MISMATCH",
+            "NO_BASELINE",
+            "SIZE_DIFFERENCE",
+        ],
+        sauceLabs: {
+            platform: platform,
+            version: "latest",
+            screenResolution: screenResolution,
+            extendedDebugging: true,
+            maxDuration: 120,
```
Comment (CITguy) [inline test/regression/runSingleTest.ts]: Do we need to comment that this value is in `seconds` and not `milliseconds`?

File: test/regression/runSingleTest.ts [hunk 13]
```diff
@@ -3,22 +3,88 @@ import {By, snap, Snappit, IConfig, WebDriver, WebElement} from "snappit-visual-
...
+        },
+        threshold: 0.1,
```
Comment (CITguy) [inline test/regression/runSingleTest.ts]: Is `threshold` needed, given we're not taking screenshots?

File: test/scripts/regressionTest.sh [hunk 14]
```diff
@@ -16,6 +16,10 @@ case $i in
     UPDATE=" -u"
     shift # past argument=value
     ;;
+    -u)
```
Comment (CITguy) [inline test/scripts/regressionTest.sh]: This is an incorrect case value. It's already defined on line 15.

File: test/regression/runSingleTest.ts (hunk 12 above)
Comment (shaleen76) [reply test/regression/runSingleTest.ts]: The max duration value is in seconds. This configuration will ensure that the sauce labs test don't exceed that time limit. 

NOTE: As per Sauce Labs, best test scripts "A test should never last more than 30 minutes and ideally should take less than five minutes."

File: test/regression/runSingleTest.ts (hunk 11 above)
Comment (shaleen76) [reply test/regression/runSingleTest.ts]: We don't need this for dom-diff. We need it only for visreg

File: test/regression/runSingleTest.ts (hunk 13 above)
Comment (shaleen76) [reply test/regression/runSingleTest.ts]: The threshold value is required.

Comment (Droogans) [reply test/regression/runSingleTest.ts]: No it isn't. Threshold is for comparing a base image to the one most recently captured for visual regression. We can remove it.

File: test/regression/runSingleTest.ts (hunk 12 above)
Comment (Droogans) [reply test/regression/runSingleTest.ts]: @CITguy it's commented already in the upstream configuration documentation.

And by "commented', I mean....the way I do it. 😬 

https://github.com/rackerlabs/snappit-visual-regression/blob/db27eb891d7c0d38abf9190d36cfcc2ad231e1fe/src/config.ts#L31

File: test/scripts/run-travis-push-or-pr [hunk 15]
```diff
@@ -2,11 +2,16 @@
...
+        npm run test:regression -- --browsers=edge
+        echo "Testing IE"
+        npm run test:regression -- --browsers=ie
         echo "Testing Chrome"
         npm run test:regression -- --browsers=chrome
-
+        echo "Testing Safari"
+        npm run test:regression -- --browsers=safari
         echo "Testing Firefox"
         npm run test:regression -- --browsers=ff
```
Comment (CITguy) [inline test/scripts/run-travis-push-or-pr]: Do we need to explicitly run regression tests for each browser individually?

Wouldn't running `npm run test:regression` (without the `--browsers` option) default to running the full suite?

Comment (shaleen76) [reply test/scripts/run-travis-push-or-pr]: npm run test:regression We can get this as well. Just how we did for components. For all components we use:  npm run test:regression --browsers=ff,chrome,safari,ie,edge

Comment (CITguy) [reply test/scripts/run-travis-push-or-pr]: Talking with @shaleen76, the reason we have separate commands is to implement a sort of "parallel" run on Saucelabs. If this is true, I'm good with it.

File: test/regression/runSingleTest.ts [hunk 16]
```diff
@@ -3,22 +3,83 @@ import {By, snap, Snappit, IConfig, WebDriver, WebElement} from "snappit-visual-
...
+            extendedDebugging: true,
+            maxDuration: 120,
+        }
     };
 
-    test(`${browser} -- ${component}`, async t => {
+    test(`${browser} on ${platform} -- ${component}`, async t => {
+        
+        if (runningSauce) {
+            config.serverUrl = `http://${process.env.SAUCE_USERNAME}:${process.env.SAUCE_ACCESS_KEY}@ondemand.saucelabs.com:80/wd/hub`;
```
Comment (Droogans) [inline test/regression/runSingleTest.ts]: What happens if you change this to `https://`?

I tried to find an example online of someone using https for this, but every single one used http.

**Comment by CITguy** [comment]
Checked "Dev LGTM" for @Droogans, given he can't submit an approval for a PR he opened and @shaleen76 has been making the updates.

**Pull request merged**